[RFC] Refactor promo sets, add missing sets as listed on Scryfall

luziferius opened the pull request:
During the conversation in issue #6180 @JayDi85 indicated that refactoring the old sets is possible, because it does not totally destroy already existing deck files. https://github.com/magefree/mage/issues/6180#issuecomment-573886961

So I started to go through all sets listed on Scryfall and add them to XMage, from old to new.
The intention is to fix all old sets and promotional releases.
- XMage should be able to download images for _all_ official MTG cards from Scryfall.
- _All_ printings of supported cards should be available within XMage.
- Edit: Fix this warning issued by the CI system: `Warning: total missing sets: 256, with missing cards: 11596`


# Progress report

## TODO
These tasks contain a bit more work than simply adding a new class describing the set.

- - [x] [Prerelease Events](https://scryfall.com/sets/ppre) contains cards until 2008, newer Prerelease promos are placed in their respective’s set’s promo cards set. Update https://github.com/magefree/mage/blob/master/Mage.Sets/src/mage/sets/PrereleaseEvents.java to reflect that. Also update the Set release date and code.
    - - [x] Updated the set code.
- - [x] Split [Arena League](https://github.com/magefree/mage/blob/master/Mage.Sets/src/mage/sets/ArenaLeague.java) into the individual sets as listed on Scryfall.
- - [x] Split [Judge Promo](https://github.com/magefree/mage/blob/master/Mage.Sets/src/mage/sets/JudgePromo.java) into the individual sets as listed on Scryfall.
- - [x] Split [Friday Night Magic](https://github.com/magefree/mage/blob/master/Mage.Sets/src/mage/sets/FridayNightMagic.java) into the individual sets as listed on Scryfall.
- - [x] Split [Magic Player Rewards](https://github.com/magefree/mage/blob/master/Mage.Sets/src/mage/sets/MagicPlayerRewards.java) into the individual sets as listed on Scryfall.
- - [x] Can the booster related settings be removed? The set has no boosters.
  https://github.com/magefree/mage/blob/68e5932d37e78b22032592bbeafe0842ad18b373/Mage.Sets/src/mage/sets/Starter2000.java#L25-L29
- - [ ] Verify Correctness of the MtgJson translation table located here:
  https://github.com/magefree/mage/blob/master/Mage.Verify/src/main/java/mage/verify/MtgJson.java
  It seems that newer JSON documents now have fully UPPER case set abbreviations.
- - [x] Make sure there are no Unicode characters in collector number strings and card names.
- - [ ] Update the sets list at https://github.com/magefree/mage/blob/master/Utils/mtg-sets-data.txt?
- - [x] Update all image downloader plugins to work with the new structure (Not necessary, as indicated in a comment below) **JayDi85: scryfall updated, other plugins are outdated and must be reworked**
- - [x] Provide a map to convert decks to the new set structure **JayDi85: no needs, xmage auto-fixes unknown set codes**
    - ~~A static HashMap that provides a lookup-table to update the user decks should be sufficient.~~
- - [x] Make sure that the Collector numbers of the [Junior Series Europe (PJSE)](https://scryfall.com/sets/pjse?order=name&as=checklist) set are not interpreted as numbers in scientific notation. These look like "1E08", which may get converted to large integers by accident.
- - [x] Build the "Supported Sets" set list in the scryfall downloader by iterating over all classes. Extract the required code from the classes instead of hard-coding it in a giant list.
- - [ ] Add all ~7 variants of [Tamiyo's Journal](https://scryfall.com/card/soi/265/tamiyos-journal) Both in SOI and PSOI.
- - [x] Add MTG Arena printings?
  If Yes: https://scryfall.com/sets/ana https://scryfall.com/sets/pana https://scryfall.com/sets/ha1
- - [x] Split the WPN and Gateway Promos?
## Added sets

All [Sets](https://scryfall.com/sets) are added up to the recently announced SLU (Secret Lair: Ultimate Edition).

### Exceptions

The [Celebration Cards (PCEL)](https://scryfall.com/sets/pcel) set is basically silver-bordered, except for two cards, which have actually functional and implementable rules text. These cards are illegal in all formats.

These non-English sets are _not_ added. Adding these will allow players to force the display of non-English card images. Therefore I opted to omit them for now.

- [Foreign Black Border (FBB)](https://scryfall.com/sets/fbb)
- [Fourth Edition Foreign Black Border (4BB)](https://scryfall.com/sets/4bb)
- [Rinascimento (RIN)](https://scryfall.com/sets/rin)
- [Renaissance (REN)](https://scryfall.com/sets/ren)
- [Redemption Program (PRED)](https://scryfall.com/sets/pred)
- [Hobby Japan Promos (PHJ)](https://scryfall.com/sets/phj)
- [Japan Junior Tournament (PJJT)](https://scryfall.com/sets/pjjt)
- [Magic Premiere Shop 2005 (PMPS)](https://scryfall.com/sets/pmps) and following
- [Salvat 2005 (PSAL)](https://scryfall.com/sets/psal)
- [Salvat 2011(PS11)](https://scryfall.com/sets/ps11)

## Notes

Please review the added classes, as they are mostly automatically generated (see below). I’ve checked them for plausibility, but I can’t guarantee that the automated process works in all cases. There are some bits I’m modifying in each class, like setting the Expansion type, booster content, and fixing all missing/wrongly generated references IntelliJ finds.

Adding all cards of a set by hand is tiresome and error prone (I tired out after 10 or so cards), so I’ve written a small script that can automatically generate Java class source files for XMage sets. It uses the card database code I’ve written for my project https://github.com/luziferius/MTGDeckConverter to build the class code. The script code used is included below for reference.
(Side note: Although the code I use is licensed under GPLv3+, the output of a GPL-licensed program, even if it is source code, is _not_ covered by the GPL, so it is OK to include the generated code files in XMage.)

```python
# Copyright (C) 2020 Thomas Hess <thomas.hess@udo.edu>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

import argparse
import datetime
import pathlib
import re
import string
import textwrap
from typing import NamedTuple, Iterable, Union, Tuple

import MTGDeckConverter.card_db.db


class Card(NamedTuple):
    name: str
    collector_number: str
    rarity: str

    def format_as_source_line(self, has_duplicates: bool) -> str:
        """Formats this card as a java source line that adds this card to the generated set."""
        package_name = "basiclands" \
            if self.name in basic_lands \
            else self.name[0].lower()
        class_name = name_to_camel_case(self.name)
        return card_line_template.substitute(
            card_name=self.name,
            collector_number=self.collector_number,
            rarity=self.rarity,
            card_package=package_name,
            card_class_name=class_name,
            # Default to assume regular basic lands with interchangeable artworks.
            # Or drop the token for non-"basic land" cards
            artwork_specifier=", NON_FULL_USE_VARIOUS" if has_duplicates else ""
        )


class Namespace(NamedTuple):
    """Mocks parsed arguments, used for type checking / hinting purposes."""
    database: pathlib.Path
    set_abbreviation: str
    output_path: pathlib.Path


def parse_args() -> Namespace:
    """Generate an argument parser and parse the command line arguments."""
    argument_parser = argparse.ArgumentParser(description="Generate XMage Sets")
    argument_parser.add_argument(
        "--database",
        action="store", type=pathlib.Path, default=pathlib.Path("./mtg.sqlite3"),
        help="Path to database storing the card data."
    )
    argument_parser.add_argument(
        "-o", "--output-path",
        action="store", type=pathlib.Path,
        help="Path to XMage Sets source Package. If given, a .java source will will be written there."
    )
    argument_parser.add_argument(
        "set_abbreviation", action="store",
        help="XMage java class to generate."
    )
    return argument_parser.parse_args()


basic_lands = ("Plains", "Island", "Swamp", "Mountain", "Forest")
alpha_re = re.compile('[^a-zA-Z]')
java_template = string.Template("""\
package mage.sets;

import mage.cards.ExpansionSet;
import mage.constants.Rarity;
import mage.constants.SetType;

/**
 * https://scryfall.com/sets/${set_abbr_lower}
 */
public class ${set_class_name} extends ExpansionSet {

    private static final ${set_class_name} instance = new ${set_class_name}();

    public static ${set_class_name} getInstance() {
        return instance;
    }

    private ${set_class_name}() {
        super("${set_name}", "${set_abbr_upper}", ExpansionSet.buildDate(${set_release_date}), SetType.);
        // this.hasBoosters = false;
        this.hasBasicLands = ${set_has_basics};

${card_list}
     }
}
""")
card_line_template = string.Template(
    'cards.add(new SetCardInfo('
    '"${card_name}", ${collector_number}, Rarity.${rarity}, mage.cards.${card_package}.${card_class_name}.class${artwork_specifier})'
    ');'
)


def validate_input_set(args: Namespace, db: MTGDeckConverter.card_db.db.CardDatabase):
    """Raise an Exception, if the given set is not known. Prevents generation of broken Java source files."""
    if not db.is_set_abbreviation_known(args.set_abbreviation.lower()):
        raise ValueError("Given set abbreviation is unknown!")


def word_to_title_case(word: str) -> str:
    if len(word) > 1:
        return word.title()
    else:
        return word


def name_to_camel_case(set_name: str) -> str:
    """Converts a string containing a name of something (Card or Set name) to a valid Java CamelCase identifier"""
    filtered_name = alpha_re.sub(" ", set_name)
    words = filtered_name.split(" ")
    
    camel_case = "".join(word_to_title_case(word) for word in words)
    return camel_case


def format_all_card_lines(args: Namespace, db: MTGDeckConverter.card_db.db.CardDatabase) -> str:
    """
    Queries the database for all files in the given set, and formats the card-adding source lines.
    Sorts the cards by the English name in ascending order.
    """
    cards_query: Iterable[Tuple[str, Union[str, int], str, int]] = db.db.execute(
        "SELECT P1.English_Name, P1.Collector_Number, P1.Rarity, Count(*) > 1 AS has_duplicates "
        "FROM Printings_View AS P1 "
        "INNER JOIN Printings_View AS P2 USING (English_Name, Abbreviation) "
        "WHERE Abbreviation = ? "
        "GROUP BY P1.English_Name, P1.Collector_Number, P1.Rarity "
        "ORDER BY P1.English_Name ASC, P1.Collector_Number ASC",
        (args.set_abbreviation.lower(),)
    )
    cards = (
        (Card(name, format_collector_number(collector_number), format_rarity(rarity)), bool(has_duplicates))
        for name, collector_number, rarity, has_duplicates in cards_query)

    card_list = "\n".join(card.format_as_source_line(has_duplicates) for (card, has_duplicates) in cards)
    return textwrap.indent(card_list, " "*8)


def format_rarity(rarity: str) -> str:
    """Converts the internal rarity name to an XMage Rarity enum member name."""
    return rarity.upper()


def format_collector_number(collector_number: Union[str, int]) -> str:
    """Collectors number may be an int or string. Convert both types to str for formatting purposes."""
    if isinstance(collector_number, int):
        return str(collector_number)
    else:
        # Put quotation marks around the collector number. Converts this into a Java string literal.
        return f'"{collector_number}"'


def format_release_date(date_str: str) -> str:
    """Format the release date in a format suitable to be put into the ExpansionSet.buildDate() constructor."""
    date = datetime.date.fromisoformat(date_str)
    # Format manually, because date.strftime() will create zero-padded numbers, which won’t work in Java code,
    # as they will be interpreted as octal literals.
    date_str = f"{date.year}, {date.month}, {date.day}"
    return date_str


def set_has_basics(args: Namespace, db: MTGDeckConverter.card_db.db.CardDatabase) -> str:
    """Returns a Java boolean literal true/false indicating if the set has basic lands."""
    # Use that the tuple string representation is a valid SQL statement for the IN operator.
    basic_lands_set = str(tuple(basic_lands))
    has_basics = bool(db.db.execute(
        "SELECT EXISTS( SELECT * "
        "FROM Printings_View "
        "WHERE Abbreviation = ? "
        f"AND English_Name IN {basic_lands_set} )",
        (args.set_abbreviation.lower(),)
    ).fetchone()[0])
    return str(has_basics).lower()


def main():
    args = parse_args()
    db = MTGDeckConverter.card_db.db.CardDatabase(args.database)
    validate_input_set(args, db)
    set_abbr_upper = args.set_abbreviation.upper()
    set_abbr_lower = args.set_abbreviation.lower()
    set_name, set_release_date = db.db.execute(
        "SELECT English_Name, Release_date "
        "FROM Card_Set "
        "WHERE Abbreviation = ?",
        (set_abbr_lower,)).fetchone()  # type: str, str
    set_release_date = format_release_date(set_release_date)
    class_name = name_to_camel_case(set_name)
    card_list = format_all_card_lines(args, db)
    has_basics = set_has_basics(args, db)
    java_class = java_template.substitute(
        set_abbr_lower=set_abbr_lower,
        set_abbr_upper=set_abbr_upper,
        set_release_date=set_release_date,
        set_name=set_name,
        set_class_name=class_name,
        card_list=card_list,
        set_has_basics=has_basics
    )
    print(java_class)
    if args.output_path is not None and args.output_path.is_dir():
        file_path = args.output_path / f"{class_name}.java"
        if file_path.exists():
            raise ValueError("Java file already exists!")
        else:
            file_path.write_text(java_class, encoding="utf-8")


if __name__ == '__main__':
    main()
```

The database has to be seeded once, using a snippet like this:
```python
import MTGDeckConverter.card_db.db 
from pathlib import Path 
db = MTGDeckConverter.card_db.db.CardDatabase(Path("./mtg.sqlite3")) 
db.populate_database() 
db.db.commit()
```

JayDi85 commented:
>1. Is it possible to add cards that do not appear in boosters to a set with boosters? (Mostly relevant for 8ED/9ED starter packs.)

If set have non booster cards then you must setup `maxCardNumberInBooster` (all cards with number after that will be ignored in booster generation). See example ThroneOfEldraine. If you can't split it by number then make two different set.
![shot_200116_194353](https://user-images.githubusercontent.com/8344157/72539303-8a052980-3898-11ea-977f-c79d2f2a2f23.png)

>2. Is it sufficient to add a .java source file to Mage.Sets/src/mage/sets/ to add basic support for a new set?

Yes, you must create one file per set. Java files loaded automaticity by client/server on startup and generates full sets/cards database. Database is basic source info about sets and cards. Moreover each set have different settings for boosters and other special code for booster generating (like special lands, partners, etc).

> 3. Should I add gold-bordered sets? (E.g. https://scryfall.com/sets/ptc) Those aren’t legal in sanctioned tournaments, but are reprints of otherwise legal cards.

If scryfall marks it as standlone set then yes -- you can add it. All sets have SetType settings. That's uses for legality check in main formats. If you setup it with `SetType.PROMOTIONAL` then only freeform game modes will be available for it.

> 4. Should NON_FULL_USE_VARIOUS to any non-full-art basic land, even if the set only has one of a kind? This is used to generate the list of available basic lands for the "add random basic lands to a deck" functionality built into the deck builder, right?

Nope. Use that settings if set contains cards with SAME name (like multiple lands or normal and bonus versions of the card). You will get error in tests (MageVerify) if something wrong. MageVerify checks many settings for cards in sets. I strongly recommends to use it (IntelliJ can run that test from IDE -- just click on mage-verify project and choose run tests).

> 5. I’m getting NPEs in the deck editor when running the client within IntelliJ and the card display pop-up window does not work. And the card image fetcher indefinitely han

If you run by IDE then check java version in project settings. It must be 1.8 (java 8) only. Newer version do not supports.

luziferius commented:
Thank you for answering.

> > 1. Is it possible to add cards that do not appear in boosters to a set with boosters? (Mostly relevant for 8ED/9ED starter packs.)
> 
> If set have non booster cards then you must setup `maxCardNumberInBooster` (all cards with number after that will be ignored in booster generation). See example ThroneOfEldraine. If you can't split it by number then make two different set.

Ok. The cards in question have collectors number `S1` to `S7`. So probably it has to remain split as it is.
https://github.com/magefree/mage/blob/master/Mage.Sets/src/mage/sets/EighthEdition.java
https://github.com/magefree/mage/blob/master/Mage.Sets/src/mage/sets/EighthEditionBox.java

> 
> > 2. Is it sufficient to add a .java source file to Mage.Sets/src/mage/sets/ to add basic support for a new set?
> 
> Yes, you must create one file per set. Java files loaded automaticity by client/server on startup and generates full sets/cards database. Database is basic source info about sets and cards. Moreover each set have different settings for boosters and other special code for booster generating (like special lands, partners, etc).

I looked at several sets (especially the newer ones) to familiarize myself with the format.

I’m adding blocks like this where appropriate:
https://github.com/magefree/mage/blob/bc899e7ea53977ed494ead237d46208481bb76c0/Mage.Sets/src/mage/sets/Battlebond.java#L20-L28

Ok. If the source files are automatically detected, I don’t have to add them to some registry file. 

> 
> > 3. Should I add gold-bordered sets? (E.g. https://scryfall.com/sets/ptc) Those aren’t legal in sanctioned tournaments, but are reprints of otherwise legal cards.
> 
> If scryfall marks it as standlone set then yes -- you can add it. All sets have SetType settings. That's uses for legality check in main formats. If you setup it with `SetType.PROMOTIONAL` then only freeform game modes will be available for it.
> 

Good to know.
What about tournament legal promotional sets?


> > 4. Should NON_FULL_USE_VARIOUS to any non-full-art basic land, even if the set only has one of a kind? This is used to generate the list of available basic lands for the "add random basic lands to a deck" functionality built into the deck builder, right?
> 
> Nope. Use that settings if set contains cards with SAME name (like multiple lands or normal and bonus versions of the card). You will get error in tests (MageVerify) if something wrong. MageVerify checks many settings for cards in sets. I strongly recommends to use it (IntelliJ can run that test from IDE -- just click on mage-verify project and choose run tests).
> 

For Example, Battlebond only has 1 artwork per basic land and still set this flag. That’s why I asked.
https://github.com/magefree/mage/blob/bc899e7ea53977ed494ead237d46208481bb76c0/Mage.Sets/src/mage/sets/Battlebond.java#L109-L111

> > 5. I’m getting NPEs in the deck editor when running the client within IntelliJ and the card display pop-up window does not work. And the card image fetcher indefinitely han
> 
> If you run by IDE then check java version in project settings. It must be 1.8 (java 8) only. Newer version do not supports.

I’ve read about it and configured my IDE accordingly when I set up the environment. It runs XMage on Java 8, but I still get those NPEs. It does find the OpenJFX libraries, which are unavailable on Java > 8, so it must be 8.

theelk801 commented:
I'm pretty sure that as long as the expansion has the right `SetType` then it won't affect legality but its cards will still be legal in the correct places based on name.

JayDi85 commented:
> Ok. The cards in question have collectors number S1 to S7. So probably it has to remain split as it is.

If scryfall split it then xmage must do same (I known there are little amount of cards in that sets, but that's easy to support, e.g. for auto-generators, checks, images downloader, etc). After add thats sets will be completed then I can modify sets search dialog to filter only basic sets (non promotional).

> What about tournament legal promotional sets?

As theelk801 above: xmage collects card names for legality checks. If any legal set contains card then you can use same card from any other set.

> For Example, Battlebond only has 1 artwork per basic land and still set this flag. That’s why I asked.

That's not critical. If one card in set then you can remove NON_FULL_USE_VARIOUS (e.g. you can remove it from battlebond's land). BTW: there are possible another settings like full art cards or special BFZ frame (that's info for M15 render mode).

> I’ve read about it and configured my IDE accordingly when I set up the environment.

Well, try to clean up:
* delete all `cards.h2.*` files from all project folder (client and server);
* delete images folder `Mage.Client\plugins\images`;
* if you want then reset all xmage settings (just delete all settings with [that instructions](https://github.com/magefree/mage/issues/4461#issuecomment-361108597))

That's error is system related. Maybe something wrong with java or paths.

theelk801 commented:
Battlebond has the tag because the automated script I used to generate the file always adds it to basics.

JayDi85 commented on Mage.Sets/src/mage/sets/AsiaPacificLandProgram.java:
```diff
@@ -18,8 +18,10 @@ public static AsiaPacificLandProgram getInstance() {
 
     private AsiaPacificLandProgram() {
-        super("Asia Pacific Land Program", "APAC", ExpansionSet.buildDate(1997, 10, 13), SetType.PROMOTIONAL);
+        super("Asia Pacific Land Program", "PALP", ExpansionSet.buildDate(1997, 10, 13), SetType.PROMOTIONAL);
```
If you change set code then all set code usage must be changed too (some download or verify source code uses set code correction for third party services):

![shot_200116_223656](https://user-images.githubusercontent.com/8344157/72552674-d5c3cd00-38b0-11ea-901e-6478f4ea15cd.png)

luziferius replied:
Thanks for notifying. Should I revert it or fix the other occurences?

JayDi85 replied:
Yes, you can replace it in other places too.

luziferius replied:
Replaced the instances. Updated the comment (first hit in your screenshot). Should be solved.

JayDi85 commented on Mage.Sets/src/mage/sets/JuniorSuperSeries.java:
```diff
@@ -0,0 +1,42 @@
+        this.hasBoosters = false;
+        this.hasBasicLands = false;
+
+        cards.add(new SetCardInfo("City of Brass", 6, Rarity.SPECIAL, mage.cards.c.CityOfBrass.class));
```
Why it special rarity? It's a rare in inscryfall: https://scryfall.com/card/psus/6/city-of-brass

luziferius replied:
Because Scryfall only categorizes into (Common, Uncommon, Rare, Mythic rare).

- https://scryfall.com/card/vma/2/time-walk is marked as Mythic, while in XMage it is `Rarity.BONUS`.
- https://scryfall.com/card/eld/266/forest is a Common on Scryfall and a `Rarity.LAND` in XMage.

JayDi85 replied:
All basic land must be LAND type. SPECIAL type uses only for two ravnika set to find special card:
![shot_200117_221601](https://user-images.githubusercontent.com/8344157/72635884-4af9d580-3977-11ea-85b8-b0a9dc0e6162.png)

I don't like SPECIAL -- it can be removed in the feature to support scryfall compatible (scryfall rarity uses real card's rarity, but official gatherer uses rarity as "booster place"). Search `public List<CardInfo> getSpecialLand()` for special cards usage in booster without SPECIAL rarity type (as example).

So:
1. Don't change rarity types in ravnika sets (it's can be next task to remove outdated SPECIAL type);
2. For all new sets add real rarity from scryfall (and LAND for basic lands), not special;
3. You can remove SPECIAL rarity from old promotional sets and make it same with scryfall;

luziferius replied:
Ok, if that's the direction, I'll do so.
And I will change all SPECIAL cards I added in the commits of thid. PR.

luziferius replied:
Replaced the SPECIAL rarity with RARE in all sets added by me (as listed on Scryfall).

I’ve checked them against Scryfall individually again, to make sure the rarity matches.
I have rebased and squashed the individual fix commits, the newly pushed commits don’t contain SPECIAL any more.

Side note: This’ll reduce my work, as I can just let my automation script run without the need to change rarity to SPECIAL any more ;) )

Should be resolved.

luziferius commented on Mage.Client/src/main/java/org/mage/plugins/card/dl/sources/ScryfallImageSupportCards.java:
```diff
@@ -20,7 +20,6 @@
             put("MPS-AKH", "mp2").
             put("MBP", "pmei").
             put("WMCQ", "pwcq").
-            put("EURO", "pelp").
```
Translation no longer needed, as XMage and Scryfall now match.

JayDi85 commented:
> ERROR 2020-01-20 14:21:23,554 java.lang.IllegalArgumentException: Set code PTC already exists.
![shot_200120_183132](https://user-images.githubusercontent.com/8344157/72734547-605b4380-3bb3-11ea-8049-980f53bb5159.png)

You can remove old cumulative PTC set from xmage and replace it by new multiple sets.

luziferius commented:
> > ERROR 2020-01-20 14:21:23,554 java.lang.IllegalArgumentException: Set code PTC already exists.
> 
> You can remove old cumulative PTC set (for pre-release events and sets) from xmage and replace it by new multiple sets.

Didn't notice the conflict and rushed the recent commits a bit...
I'll check everything locally when I think it's ready to request another round of reviews.

Those cummulative sets are on my TODO list. (See the check list I added to the initial PR description.)

Oh. And I found a set registry (kind of) in the plugins. Theres a list of all supported Scryfall sets that needs to be update with all sets added.

JayDi85 commented:
You can find error logs by github, without local run:

![shot_200120_232551](https://user-images.githubusercontent.com/8344157/72752725-69143f80-3bdc-11ea-94f4-306a6ffb4df9.png)

![shot_200120_232618](https://user-images.githubusercontent.com/8344157/72752739-70d3e400-3bdc-11ea-9b41-28c54b601f1b.png)

luziferius commented:
I’ve seen the CI links, but normally I run them locally, because it’s faster.

But now something else:
I’m currently splitting the Arena League promos and a search for `ARENA` led me to this:
https://github.com/magefree/mage/blob/50000205920e8856c437804e6e5664bc01ff1ded/Mage.Verify/src/main/java/mage/verify/MtgJson.java#L28-L50
The map translates `pARL` to `ARENA`. Now that `ARENA` is removed, I wanted to verify how MtgJson handles the other sets (`PAL99`-`PAL06`). According to that block, the initial `p` is lower case.
Then I downloaded the JSON document from `https://mtgjson.com/json/AllSets.json.zip` and loaded it into Python for inspection. I noticed that all sets are fully UPPER case, thus it seems that the translation is partly outdated. 

I’ll add the verification of that code to the TODO list above. Maybe most of those items can be removed later.

```python
>>> import json, pprint, pathlib
>>> mtg_data=json.loads(pathlib.Path("./AllPrintings.json").read_text())

for set_ in mtg_data.keys():
...     if set_.startswith("p"): 
...         print(set_)

pprint.PrettyPrinter(compact=True).pprint(list(mtg_data.keys()))                                                                                                                                                                               
['10E', '2ED', '3ED', '4BB', '4ED', '5DN', '5ED', '6ED', '7ED', '8ED', '9ED',
 'A25', 'AER', 'AKH', 'ALA', 'ALL', 'AMH1', 'ANA', 'APC', 'ARB', 'ARC', 'ARN',
 'ATH', 'ATQ', 'AVR', 'BBD', 'BFZ', 'BNG', 'BOK', 'BRB', 'BTD', 'C13', 'C14',
 'C15', 'C16', 'C17', 'C18', 'C19', 'CED', 'CEI', 'CHK', 'CHR', 'CM1', 'CM2',
 'CMA', 'CMB1', 'CMD', 'CN2', 'CNS', 'CON', 'CP1', 'CP2', 'CP3', 'CSP', 'CST',
 'DD1', 'DD2', 'DDC', 'DDD', 'DDE', 'DDF', 'DDG', 'DDH', 'DDI', 'DDJ', 'DDK',
 'DDL', 'DDM', 'DDN', 'DDO', 'DDP', 'DDQ', 'DDR', 'DDS', 'DDT', 'DDU', 'DGM',
 'DIS', 'DKA', 'DKM', 'DOM', 'DPA', 'DRB', 'DRK', 'DST', 'DTK', 'DVD', 'E01',
 'E02', 'ELD', 'EMA', 'EMN', 'EVE', 'EVG', 'EXO', 'EXP', 'F01', 'F02', 'F03',
 'F04', 'F05', 'F06', 'F07', 'F08', 'F09', 'F10', 'F11', 'F12', 'F13', 'F14',
 'F15', 'F16', 'F17', 'F18', 'FBB', 'FEM', 'FNM', 'FRF', 'FUT', 'G00', 'G01',
 'G02', 'G03', 'G04', 'G05', 'G06', 'G07', 'G08', 'G09', 'G10', 'G11', 'G17',
 'G18', 'G99', 'GK1', 'GK2', 'GN2', 'GNT', 'GPT', 'GRN', 'GS1', 'GTC', 'GVL',
 'H09', 'H17', 'HA1', 'HHO', 'HML', 'HOP', 'HOU', 'HTR', 'HTR17', 'HTR18',
 'ICE', 'IMA', 'INV', 'ISD', 'ITP', 'J12', 'J13', 'J14', 'J15', 'J16', 'J17',
 'J18', 'J19', 'J20', 'JGP', 'JOU', 'JUD', 'JVC', 'KLD', 'KTK', 'L12', 'L13',
 'L14', 'L15', 'L16', 'L17', 'LEA', 'LEB', 'LEG', 'LGN', 'LRW', 'M10', 'M11',
 'M12', 'M13', 'M14', 'M15', 'M19', 'M20', 'MB1', 'MBS', 'MD1', 'ME1', 'ME2',
 'ME3', 'ME4', 'MED', 'MGB', 'MH1', 'MIR', 'MM2', 'MM3', 'MMA', 'MMQ', 'MOR',
 'MP2', 'MPR', 'MPS', 'MRD', 'NEM', 'NPH', 'OARC', 'OC13', 'OC14', 'OC15',
 'OC16', 'OC17', 'OC18', 'OC19', 'OCM1', 'OCMD', 'ODY', 'OE01', 'OGW', 'OHOP',
 'OLGC', 'ONS', 'OPC2', 'OPCA', 'ORI', 'OVNT', 'P02', 'P03', 'P04', 'P05',
 'P06', 'P07', 'P08', 'P09', 'P10', 'P10E', 'P11', 'P15A', 'P2HG', 'PAER',
 'PAKH', 'PAL00', 'PAL01', 'PAL02', 'PAL03', 'PAL04', 'PAL05', 'PAL06', 'PAL99',
 'PALP', 'PANA', 'PARC', 'PARL', 'PAVR', 'PBBD', 'PBFZ', 'PBNG', 'PBOK', 'PC2',
 'PCA', 'PCEL', 'PCMD', 'PCMP', 'PCY', 'PD2', 'PD3', 'PDGM', 'PDKA', 'PDOM',
 'PDP10', 'PDP11', 'PDP12', 'PDP13', 'PDP14', 'PDRC', 'PDTK', 'PDTP', 'PELD',
 'PELP', 'PEMN', 'PF19', 'PF20', 'PFRF', 'PG07', 'PG08', 'PGPX', 'PGRN', 'PGRU',
 'PGTC', 'PGTW', 'PHEL', 'PHJ', 'PHOP', 'PHOU', 'PHPR', 'PHUK', 'PI13', 'PI14',
 'PIDW', 'PISD', 'PJAS', 'PJJT', 'PJOU', 'PJSE', 'PKLD', 'PKTK', 'PLC', 'PLGM',
 'PLNY', 'PLPA', 'PLS', 'PM10', 'PM11', 'PM12', 'PM13', 'PM14', 'PM15', 'PM19',
 'PM20', 'PMBS', 'PMEI', 'PMH1', 'PMOA', 'PMPS', 'PMPS06', 'PMPS07', 'PMPS08',
 'PMPS09', 'PMPS10', 'PMPS11', 'PNAT', 'PNPH', 'POGW', 'POR', 'PORI', 'PPC1',
 'PPOD', 'PPP1', 'PPRE', 'PPRO', 'PR2', 'PRED', 'PREL', 'PRES', 'PRIX', 'PRM',
 'PRNA', 'PROE', 'PRTR', 'PRW2', 'PRWK', 'PS11', 'PS14', 'PS15', 'PS16', 'PS17',
 'PS18', 'PS19', 'PSAL', 'PSDC', 'PSLD', 'PSOI', 'PSOM', 'PSS1', 'PSS2', 'PSS3',
 'PSUM', 'PSUS', 'PTC', 'PTG', 'PTHS', 'PTK', 'PTKDF', 'PUMA', 'PURL', 'PUST',
 'PVAN', 'PWAR', 'PWCQ', 'PWOR', 'PWOS', 'PWP09', 'PWP10', 'PWP11', 'PWP12',
 'PWPN', 'PWWK', 'PXLN', 'PXTC', 'PZ1', 'PZ2', 'PZEN', 'RAV', 'REN', 'RIN',
 'RIX', 'RNA', 'ROE', 'RQS', 'RTR', 'S00', 'S99', 'SCG', 'SHM', 'SLD', 'SOI',
 'SOK', 'SOM', 'SS1', 'SS2', 'STH', 'SUM', 'TBTH', 'TD0', 'TD2', 'TDAG', 'TFTH',
 'THB', 'THP1', 'THP2', 'THP3', 'THS', 'TMP', 'TOR', 'TPR', 'TSB', 'TSP', 'UDS',
 'UGIN', 'UGL', 'ULG', 'UMA', 'UNH', 'USG', 'UST', 'V09', 'V10', 'V11', 'V12',
 'V13', 'V14', 'V15', 'V16', 'V17', 'VIS', 'VMA', 'W16', 'W17', 'WAR', 'WC00',
 'WC01', 'WC02', 'WC03', 'WC04', 'WC97', 'WC98', 'WC99', 'WTH', 'WWK', 'XLN',
 'ZEN']
```

luziferius commented:
About the CI failure with the [Magic Online Promos (PRM)](https://scryfall.com/sets/prm?order=name) set: `mage.verify.VerifyCardDataTest` fails with this error:
```
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Damnation - 31417
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Day of Judgment - 37873
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Disenchant - 31397
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Doom Blade - 37855
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Fireball - 31485
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Giant Growth - 35824
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Lightning Bolt - 36224
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Negate - 35138
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Terminate - 35136
error, founded double card names, but UsesVariousArt = false (missing NON_FULL_USE_VARIOUS, etc): PRM - Magic Online Promos - Terror - 31483
```

All errors are of the same kind:
The set has both a full-art version and a non-full-art version of the given cards above.

Example query for Damnation:
https://scryfall.com/search?q=set%3Aprm+unique%3Aprints+Damnation&order=name

I defined a custom `CardGraphicInfo` instance:
https://github.com/magefree/mage/blob/d3b75b6196566813c43c89a8f061e9dd8916ca01/Mage.Sets/src/mage/sets/MagicOnlinePromos.java#L20
And used it to mark the appropriate full-art cards:
https://github.com/magefree/mage/blob/d3b75b6196566813c43c89a8f061e9dd8916ca01/Mage.Sets/src/mage/sets/MagicOnlinePromos.java#L233-L234

JayDi85 commented:
You need to move new graphic info object to constructor:

![shot_200204_003208](https://user-images.githubusercontent.com/8344157/73688626-e0ba9200-46e5-11ea-8ebc-0ffa16d60af6.png)

JayDi85 commented on Mage.Sets/src/mage/sets/JudgeGiftCards2014.java:
```diff
@@ -0,0 +1,38 @@
+        cards.add(new SetCardInfo("Oloro, Ageless Ascetic", 9, Rarity.RARE, mage.cards.o.OloroAgelessAscetic.class));
+        cards.add(new SetCardInfo("Plains", "1★", Rarity.LAND, mage.cards.basiclands.Plains.class));
+        cards.add(new SetCardInfo("Riku of Two Reflections", 3, Rarity.MYTHIC, mage.cards.r.RikuOfTwoReflections.class));
+        cards.add(new SetCardInfo("Swamp", "3★", Rarity.LAND, mage.cards.basiclands.Swamp.class));
```
Don't use unicode symbols in card numbers or names, use only ascii (there are some regexp with save/load deck files -- it can fail with that card numbers chars).

luziferius replied:
Ok, This was copied from Scryfall as-is.
Then I’ll remove the ★ and add a special download rule.

luziferius replied:
This may be a nice addition to the card data verification test.
It should be easy to test the card names and numbers against the RE `[a-zA-Z0-9'_"]+` or similar.

luziferius replied:
Replaced the ★.

luziferius commented on Mage.Sets/src/mage/sets/ArenaLeague1999.java:
```diff
@@ -0,0 +1,34 @@
+
+        cards.add(new SetCardInfo("Forest", 1, Rarity.LAND, mage.cards.basiclands.Forest.class));
+        cards.add(new SetCardInfo("Island", 3, Rarity.LAND, mage.cards.basiclands.Island.class, NON_FULL_USE_VARIOUS));
+        cards.add(new SetCardInfo("Island", "3†", Rarity.LAND, mage.cards.basiclands.Island.class, NON_FULL_USE_VARIOUS));
```
Non-ASCII collector number.

luziferius commented on Mage.Sets/src/mage/sets/JuniorApacSeries.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * https://scryfall.com/sets/pjas
+ */
+public class JuniorApacSeries extends ExpansionSet {
```
Needs to be `JuniorAPACSeries`. Already corrected locally and will be fixed on next push.

luziferius commented:
What is the stance on non-English-only cards, i.e. printings that are only available in languages other than English?

Examples:
- [Rinascimento (RIN)](https://scryfall.com/sets/rin) is a full set not available in English
- [Pro Tour Promos (PPRO)](https://scryfall.com/sets/ppro?order=name) and several other promotional sets and special printings contain _some_ non-English cards. (Example: https://scryfall.com/card/ppro/2019-3/ゴルガリの女王、ヴラスカ)

Should I omit those cards and sets or add them?


If added, players can use those printings to force the display of non-English cards on other players, by choosing for example a Japanese-only printing.

JayDi85 commented:
Nope, do not add non english sets.

JayDi85 commented:
Found some TODOs in start topic:

> Can the booster related settings be removed? The set has no boosters

Yes, if set don't have boosters then you don't need that settings.

> Update the sets list at https://github.com/magefree/mage/blob/master/Utils/mtg-sets-data.txt?

If you can then yes. But it's not necessary. That's file uses for card code template generating and other perl scripts.

> Update all image downloader plugins to work with the new structure
1. What's changes are you talking about?

> Provide a map to convert decks to the new set structure
> A static HashMap that provides a lookup-table to update the user decks should be sufficient.
2. Are you talking about auto-convert deck files in xmage release folder, user's computer or something else?

luziferius commented:
>     1. What's changes are you talking about?

The "catch-all" sets, e.g. FNM or Judge Promo releases that I split into each year. I assume that some image sources do not have that split. So to continue to support them, the individual sets have to be merged again for download purposes. So I’m not quite sure whether splitting some of the "group-sets", like FNM, Arena League Promos, etc is a good idea.

For example, this set: 
https://github.com/magefree/mage/pull/6190/commits/d28ab43951d1fa9f7a85328267438316823ccf9d

> 
> > Provide a map to convert decks to the new set structure
> > A static HashMap that provides a lookup-table to update the user decks should be sufficient.
> 
>     3. Are you talking about auto-convert deck files in xmage release folder, user's computer or something else?

Semi-automatically convert decks on user’s computer. I envision to add a button to the deck editor that allows the user to select deck files. It uses a Map that contains a mapping from (previous set code + number) to (new set code + number) for all changed printings and use this map to update all selected deck files.

It is meant to provide the users a smooth transition, so that they don’t have to fix their bling decks manually. This is an optional stretch-goal. I wouldn’t like it if all my carefully crafted bling-decks got wiped with the inclusion of this PR. So I envisioned a method to keep user’s decks.

JayDi85 commented:
Scryfall is main images source. Other sources updated last time 1+ years ago. Its not necessary to support it.

JayDi85 commented:
Also you don't need to convert card numbers. I can add priority on card name + set code instead set code + card number. E.g. card number will be used only for cards with multiple images in set. Maybe that can simplified convert routine (only set code replace map needed).

JayDi85 commented:
BTW sets must be split as scryfall. If another images source uses combined source then just add refs from split sets to one single set code in "set code replacement" map. Thats must work fine.

luziferius commented:
@JayDi85 I think there is some issue with unwanted non-determinism somewhere in the execution engine. Please look at the CI results above: There is one CI build failure. It fails because a unrelated test failed. https://travis-ci.org/github/magefree/mage/builds/662818905#L11968
The affected commit didn’t even touch any card implementation. And the next commit was building fine. Again without me touching any card implementation.

JayDi85 commented:
Yes, it's a random fail from ThousandYearStormTest. I re-launched tests and all fine.

luziferius commented:
All sets up to the announced Secret Lair are added, except some excluded sets.

What remains are further clean-up tasks in the Scryfall image downloader and verifying that all images can be fetched.

I think the CI fails because I touched a set that contains a Counterspell, which broke the test.
I’ll look into it tomorrow, or when I continue to work on this.

luziferius commented:
I’ve implemented the Mystery Booster Retail release, including accurate booster generation that takes the exact print sheets into account.
So Mystery Booster draft should be a nice new feature once this PR is finished up.

And I fixed some outstanding issues. But there is still work to do

luziferius commented:
I consider this to be mostly done.

- Almost all promotional sets are added.( No non-English cards are added.)
- The Scryfall downloader now fetches all images of all sets without issues.
- Removed quite some workarounds in the Scryfall downloader
- The downloader now shows a better total size approximation

Remaining tasks:

- [ ] Support all Tokens

JayDi85 commented:
It will be merged after the next hotfix release.

luziferius commented:
Take your time. This is a rather large change, that will affect quite some decks. I hope that the disruptions caused by this won't be too large.

luziferius commented:
Please link these issues that got fixed by this PR (to close them on merge):
- #1915 (Wastes now have both full-art and regular frame versions available)
- #2301 (The Judge reward basic land cards are available.)

Zerrisx commented:
Looks like this will also address #5021.  Any timeline on when we plan to merge this?  It's been sitting for a while.

luziferius commented:
It’ll need some conflict resolution, but most should be pretty easy to solve.

A Linux distribution update on my side made it impossible to run XMage inside the IDE (it finally made the old Java 8 OpenJFX completely incompatible). I’m waiting on OpenJDK 11 compatibility ATM.

Zerrisx commented:
Would it be worth having another developer take it over to fix up the last few conflicts and merge it for you, or should we leave it waiting for OpenJDK 11 Compatibility?

luziferius commented:
I thought of a backwards compatibility issue with this:

This PR removes and alters some sets. This will be incompatible with previously downloaded images. The removed sets won’t harm much, as their image archive will just linger on disk forever without being used.
But it has to be checked before merging that the updated sets will work correctly with the old image data or if they start to show wrong images or crash the client at worst.

JayDi85 commented:
Yes, I know. It will be in what's new docs (about images clean up and/or re-download recommendations). 

It's ok if some old promo sets will be stores in zips/folders forever (until images folder clean up).
